xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R160000001XWIKI21738DataMigration.java

```diff
@@ -0,0 +1,201 @@
+import org.xwiki.notifications.NotificationException;
+import org.xwiki.notifications.filters.NotificationFilterPreference;
+import org.xwiki.notifications.filters.internal.DefaultNotificationFilterPreference;
+import org.xwiki.notifications.filters.internal.NotificationFilterPreferenceStore;
+import org.xwiki.query.Query;
+import org.xwiki.query.QueryException;
+import org.xwiki.query.QueryManager;
+import org.xwiki.wiki.descriptor.WikiDescriptorManager;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.store.XWikiHibernateStore;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Migrate filters to put them in the same DB than where the users are located.
+ *
+ * @version $Id$
+ * @since 16.0.0RC1
+ */
+@Component
+@Named("R160000001XWIKI21738")
+@Singleton
+public class R160000001XWIKI21738DataMigration extends AbstractHibernateDataMigration
+{
+    private static final int BATCH_SIZE = 100;
+
+    private static final String SEARCH_FILTERS_STATEMENT = "select nfp "
+        + "from DefaultNotificationFilterPreference nfp "
+        + "where nfp.owner not like :ownerLike and nfp.owner <> :mainWiki "
+        + "order by nfp.owner, nfp.internalId";
+
+    private static final String DELETE_FILTER_STATEMENT = "delete from DefaultNotificationFilterPreference nfp "
+        + "where nfp.internalId IN (:listIds)";
+
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    @Inject
+    private QueryManager queryManager;
+
+    @Inject
+    private NotificationFilterPreferenceStore filterPreferenceStore;
+
+    @Inject
+    @Named("relative")
+    private EntityReferenceResolver<String> entityReferenceResolver;
+
+    @Inject
+    private EntityReferenceSerializer<String> entityReferenceSerializer;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public String getDescription()
+    {
+        return "Migrate filters next to users.";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(160000001);
+    }
+
+    @Override
+    protected void hibernateMigrate() throws DataMigrationException, XWikiException
+    {
+        // This migration only needs to be performed on main wiki: if we have filters on local wikis it's because
+        // local filters was enabled and we don't support switching mainstore / localstore.
+        // Migration steps:
+        //   - Retrieve filters from main wiki where owner belongs to a subwiki
+        //   - Store the filters on the subwiki DB using standard storage component
+        //   - Remove the filters from main DB
+
+        String mainWikiId = this.wikiDescriptorManager.getMainWikiId();
+
+        if (Objects.equals(this.wikiDescriptorManager.getCurrentWikiId(), mainWikiId)) {
```

Comment:
Doing that if is actually #shouldExecute job.

Reply:
We're doing same check [in another migration](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R140401000XWIKI15460DataMigration.java#L140) with this comment: 
```
// This check cannot be done in #shouldExecute because possibly missing columns are not yet added to the 
        // database.
```

Reply:
It's not the same check, no. Here you are only checking if you are in the main wiki, and you can perfectly do that in #shouldExecute.